File: android/app/src/main/java/com/example/jamboree/settings/AppSettings.kt
package com.example.ubapp.settings

import android.content.Context

/**
 * User-facing app preferences, persisted in `SharedPreferences`.
 *
 * - `hostName` is the display name this device uses when it hosts a
 *   browser-tier game (the player id is still `host`).
 * - `diagnosticsEnabled` is a developer toggle that reveals the on-screen
 *   host connection log in the hosting screen.
 */
object AppSettings {
    private const val PREFS = "ubapp.settings"
    private const val KEY_HOST_NAME = "hostName"
    private const val KEY_DIAGNOSTICS = "diagnosticsEnabled"

    @Volatile private var cachedDiagnostics = false

    private fun prefs(ctx: Context) =
        ctx.getSharedPreferences(PREFS, Context.MODE_PRIVATE)

    /** Trimmed host name, never empty. */
    fun hostName(ctx: Context): String {
        val raw = prefs(ctx).getString(KEY_HOST_NAME, null)?.trim().orEmpty()
        return raw.ifEmpty { "Host" }
    }

    fun setHostName(ctx: Context, value: String) {
        prefs(ctx).edit().putString(KEY_HOST_NAME, value).apply()
    }

    fun diagnosticsEnabled(ctx: Context): Boolean {
        val v = prefs(ctx).getBoolean(KEY_DIAGNOSTICS, false)
        cachedDiagnostics = v
        return v
    }

    fun setDiagnostics(ctx: Context, value: Boolean) {
        cachedDiagnostics = value
        prefs(ctx).edit().putBoolean(KEY_DIAGNOSTICS, value).apply()
    }

    /** Last known diagnostics state; safe to read without a [Context] (used
     *  by `HostServer` logging on its socket threads). Seeded via
     *  [diagnosticsEnabled] / [setDiagnostics]. */
    val diagnosticsOn: Boolean get() = cachedDiagnostics
}

File: ios/Jamboree/Settings/AppSettings.swift
import SwiftUI

/// User-facing app preferences, persisted in `UserDefaults`.
///
/// - `hostName` is the display name this device uses when it hosts a
///   browser-tier game (the player id is still `host`).
/// - `diagnosticsEnabled` is a developer toggle that reveals the on-screen
///   host connection log in the hosting screen.
@MainActor
final class AppSettings: ObservableObject {
    static let shared = AppSettings()

    nonisolated static let hostNameKey = "jamboree.hostName"
    nonisolated static let diagnosticsKey = "jamboree.diagnosticsEnabled"

    @Published var hostName: String {
        didSet { UserDefaults.standard.set(hostName, forKey: Self.hostNameKey) }
    }

    @Published var diagnosticsEnabled: Bool {
        didSet { UserDefaults.standard.set(diagnosticsEnabled, forKey: Self.diagnosticsKey) }
    }

    private init() {
        let stored = UserDefaults.standard.string(forKey: Self.hostNameKey) ?? ""
        hostName = stored.isEmpty ? "Host" : stored
        diagnosticsEnabled = UserDefaults.standard.bool(forKey: Self.diagnosticsKey)
    }

    /// Trimmed host name, never empty. Safe to read from any actor — used by
    /// game views when constructing their server.
    nonisolated static var currentHostName: String {
        let raw = UserDefaults.standard.string(forKey: hostNameKey) ?? ""
        let trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        return trimmed.isEmpty ? "Host" : trimmed
    }

    /// Whether the developer diagnostics log is enabled. Safe from any actor.
    nonisolated static var diagnosticsOn: Bool {
        UserDefaults.standard.bool(forKey: diagnosticsKey)
    }
}
